Add vcpkg.json and related files for building on Windows (#15510)
luanti-org/luanti: doc/compiling/windows.md

# Compiling on Windows using MSVC

In this guide, we use `path/to/minetest` to refer to the repository. Minetest was the old name of Luanti, and as of this writing, the repo has not been renamed.

## 1. Machine requirements

Luanti is compiled via Visual Studio and CMake, and version control is managed with Git.

-   [Visual Studio 2015 or newer](https://visualstudio.microsoft.com) including "Desktop development with C++" (includes CMake CLI but not CMake GUI)
-   [Git](https://git-scm.com/downloads)

### 1.1. Visual Studio (VS) with C++

1. Install VS from [visualstudio.microsoft.com](https://visualstudio.microsoft.com)
1. In the VS installer, select "Desktop development with C++":

    ![VS installer showing desktop development with C++ selected](assets/vs-installer.png)

> 1. Confirm the installation.

Reviewer: Why are all items numbered as 1?  It seems to be displayed properly as 1,2,3 in GitHub markdown preview but still.

Reviewer 2: markdown has automatic numbering

Author: Yes, I use auto-numbering as it's easier to move or insert items later. Resolving

## 1. Machine requirements

Luanti is compiled via Visual Studio and CMake, and version control is managed with Git.

-   [Visual Studio 2015 or newer](https://visualstudio.microsoft.com) including "Desktop development with C++" (includes CMake CLI but not CMake GUI)
-   [Git](https://git-scm.com/downloads)

### 1.1. Visual Studio (VS) with C++

1. Install VS from [visualstudio.microsoft.com](https://visualstudio.microsoft.com)
1. In the VS installer, select "Desktop development with C++":

    ![VS installer showing desktop development with C++ selected](assets/vs-installer.png)

1. Confirm the installation.

This will install the C++ compiler used in later steps. VS is also the recommended IDE for Luanti.

### 1.2. CMake

The CMake GUI is a great tool to manage first-time setup of Luanti. It's not required, but recommended for anyone unfamiliar with CMake or its CLI. CMake itself comes bundled with Visual Studio, but the CMake GUI does not, and must be installed separately.

Install from [cmake.org/download](https://cmake.org/download/). Once installed, you should be able to run `cmake-gui` from the start menu:

> ![cmake-gui in Windows start menu shows app result](./assets/cmake-gui-search.png)

Reviewer 2: I think we can assume the user knows how to use the start menu

1. Edit your PATH environment variable to include the `/path/to/bin` (but not `cmake.exe`).

    1. Start menu > Search "env" > Edit the system environment variables > Environment Variables... > Path > Edit... > New > Paste the path from step 1

    ![Path variable with cmake added](./assets/path.png)

    ![path ends with \bin, not \cmake.exe](./assets/path-bin.png)

1. Restart your terminal
1. Try `(Get-Command cmake).Path` in PowerShell or `which cmake` in Bash to ensure CMake can be executed from the terminal.

## 2. Install Luanti dependencies

Once VS and CMake are ready to use, you can install Luanti's dependencies. Luanti requires several C++ packages to function, which are installed by CMake and compiled by a C++ compiler bundled with VS. CMake uses a tool called vcpkg behind the scenes, which you may see referenced elsewhere in this project.

### 2.1. Optional dependencies

Some optional dependencies are recommended for advanced development. You can add them to `vcpkg.json` to install them along with the required dependencies.

> -   `curl` is highly recommended, as it's required to read the serverlist; `curl[winssl]` is required to use the content store.

Reviewer 2: ```suggestion
-   `curl[winssl]` is highly recommended, as it's required to use the server list and content store.
```

# Compiling on Windows using MSVC

If you're just creating mods or games with Luanti, you do not need to compile Luanti. Instead, follow Ruben Wardy's [Luanti Modding Book](https://rubenwardy.gitlab.io/minetest_modding_book) to get started modding. Compiling Luanti is only required if you plan to modify the Luanti engine itself.

## Requirements

-   [Visual Studio 2015 or newer](https://visualstudio.microsoft.com) including "Desktop development with C++"
-   [CMake](https://cmake.org/download/)
-   [vcpkg](https://github.com/Microsoft/vcpkg) (included with Visual Studio)
-   [Git](https://git-scm.com/downloads)

### Visual Studio with C++

1. Install Visual Studio (VS) from [visualstudio.microsoft.com](https://visualstudio.microsoft.com)
1. In the VS installer, select "Desktop development with C++":

    ![VS installer showing desktop development with C++ selected](assets/vs-installer.png)

1. Confirm the installation.

This will install the C++ compiler used in later steps. VS is also the recommended IDE for Luanti.

> You may notice that the C++ tools include CMake, but is should also be installed separately for compatibility with Luanti.

Author: ```suggestion
You may notice that the C++ tools include CMake, but it should also be installed separately for compatibility with Luanti.
```

> You may notice that the C++ tools include CMake, but it should also be installed separately for compatibility with Luanti.

Author: cmake-gui wasn't included in my Start menu by default, I bet I can find it though, let me see if I can simplify the docs accordingly

> You may notice that the C++ tools include CMake, but it should also be installed separately for compatibility with Luanti.

Author: Updated with instructions on how to add CMake CLI from VS to path and use that instead of GUI :)

-   `curl` is highly recommended, as it's required to read the serverlist; `curl[winssl]` is required to use the content store.
-   `openal-soft`, `libvorbis` and `libogg` are optional, but required to use sound.
-   `luajit` is optional, it replaces the integrated Lua interpreter with a faster just-in-time interpreter.
-   `gmp` and `jsoncpp` are optional, otherwise the bundled versions will be compiled
-   `gettext` is optional, but required to use translations.

### Install dependencies

1. Start up the CMake GUI (Win > search "cmake-gui" > open)
1. Select **Browse Source...** and select `path/to/minetest` (where you've cloned the repo)
1. Select **Browse Build...** and select `path/to/minetest/build` (a new folder that CMake will prompt to create)
1. Select **Configure**
1. Choose the right Visual Studio version and target platform. Currently, Luanti uses Visual Studio 16 2019, but newer VS versions should work as well. The VS version has to match the version of the installed dependencies.
1. Choose **Specify toolchain file for cross-compiling**
1. Click **Next**

> 1. Select the vcpkg toolchain file, e.g. `C:/Program Files/Microsoft Visual Studio/2022/Community/VC/vcpkg/scripts/buildsystems/vcpkg.cmake`. Save this value for later.

Author: ```suggestion
1. Select the vcpkg toolchain file, e.g. `C:/Program Files/Microsoft Visual Studio/2022/Community/VC/vcpkg/scripts/buildsystems/vcpkg.cmake`.
```

1. Click Finish
1. Wait until CMake generates the cache file (this may take about 10-30 minutes, depending on your device)
1. If there are any errors, solve them and hit **Configure**
1. Click **Generate**
1. Click **Open Project**

## Compile Luanti

There are two ways to compile Luanti: via Visual Studio or via CLI.

### Compile in Visual Studio

14. Compile Luanti inside Visual Studio.
    -   If you get "Unable to start program '...\x64\Debug\ALL_BUILD'. Access is denied", try compiling via the CLI instead.

### Compile via CLI

> While in the `path/to/minetest` folder, run the following command:

Reviewer 3: Maybe these sections should also mention that the Install Dependencies step must be completed before these will work.

Author: Hmm, I figured folks would go top-down, would numbering the headings help?

Reviewer 3: Maybe it's clear enough. I'm just bringing up the discussion point for consideration, since I can imagine a scenario where someone skips to the part that says "Compile".

Author: Numbered the headings and added a lead to each one clarifying that the previous step is required

# Compiling on Windows using MSVC

> In this guide, we use `path/to/minetest` to refer to the repository. Minetest was the old name of Luanti, and as of this writing, the repo has not been renamed.

Reviewer 2: not correct anymore :)

## 2. Install Luanti dependencies

Once VS and CMake are ready to use, you can install Luanti's dependencies. Luanti requires several C++ packages to function, which are installed by CMake and compiled by a C++ compiler bundled with VS. CMake uses a tool called vcpkg behind the scenes, which you may see referenced elsewhere in this project.

### 2.1. Optional dependencies

Some optional dependencies are recommended for advanced development. You can add them to `vcpkg.json` to install them along with the required dependencies.

-   `curl[winssl]` is highly recommended, as it's required to use the server list and content store.
-   `openal-soft`, `libvorbis` and `libogg` are optional, but required to use sound.
-   `luajit` is optional, it replaces the integrated Lua interpreter with a faster just-in-time interpreter.
-   `gmp` and `jsoncpp` are optional, otherwise the bundled versions will be compiled
-   `gettext` is optional, but required to use translations.

### 2.2. Install dependencies

You can install dependencies via CLI or via CMake GUI, whichever your prefer.

> To install dependencies via CLI, you can run this program in PowerShell:

Reviewer 2: ```suggestion
To install dependencies via CLI, run the following commands in PowerShell:
```

```powershell
# Run this while in `path/to/luanti`
$vs="Visual Studio 17 2022" # or "Visual Studio 16 2019", whatever matches your machine
$toolchain_file="C:/Program Files/Microsoft Visual Studio/2022/Community/VC/vcpkg/scripts/buildsystems/vcpkg.cmake" # or wherever you installed vcpkg.cmake manually
mkdir build

> cmake -S . -B build -G "$vs" -DCMAKE_TOOLCHAIN_FILE="$toolchain_file" -DCMAKE_BUILD_TYPE=Release -DENABLE_CURSES=OFF

Reviewer 2: running a configure installs the dependencies? that is a bit confusing even if true.
any reason not to continue recommending the `vcpkg` call like before?

Author: The `vcpkg` call is unnecessary and overcomplicates the build process, in short.

## Requirements

-   [Visual Studio 2015 or newer](https://visualstudio.microsoft.com)
-   [CMake](https://cmake.org/download/)
-   [vcpkg](https://github.com/Microsoft/vcpkg)
-   [Git](https://git-scm.com/downloads)

## Compiling and installing the dependencies

> It is highly recommended to use vcpkg as package manager, installed in a location without spaces in the path for compatibility with CMake.

Reviewer 3: ```suggestion
It is highly recommended to use vcpkg to manage dependencies, installed at a path without spaces for compatibility with the Luanti build configuration.
```

Reviewer 3: CMake handles spaces fine. It's our fault spaces don't work, although anybody using spaces should know better than to test us.

Reviewer 4: "path without spaces *in it*" sounds a little better to me

Author: ```suggestion
It is highly recommended to use vcpkg to manage dependencies, installed at a path without spaces in it for compatibility with the Luanti build configuration.
```

```

> "C:\Program Filesvcpkg\vcpkg.exe" # bad, may have issues with CMake

Reviewer 3: ```suggestion
"C:\Program Filesvcpkg\vcpkg.exe" # bad, may have issues
```

"C:\vcpkg\vcpkg.exe" # good, no spaces means fewer problems
```

If spaces are present, you may see errors like:

```
libtool:   error: 'Files/vcpkg/buildtrees/libiconv/x64-windows-dbg/lib/libcharset.la' is not a directory

> ```

Reviewer 3: I like the example. Great documentation.

After you successfully built vcpkg you can easily install the required libraries:

> ```powershell

Reviewer 3: CMake will also do installation when you invoke it, so I believe this step is technically unnecessary.

Author: Could you clarify what "this step" is? Maybe it can be removed in a separate PR? I'm personally re-building Luanti on a new machine now, I can try doing it without "this step", once I know what it is ;)

Reviewer 3: Running the vcpkg command to install dependencies. CMake will run it for you when you do a configuration.

Reviewer 3: I think it's fine to leave it here. It doesn't cause any issues. It's just something to keep in mind if you weren't aware of it, since I think fewer steps here likely means less that can go wrong.

vcpkg install zlib zstd curl[winssl] openal-soft libvorbis libogg libjpeg-turbo sqlite3 freetype luajit gmp jsoncpp gettext[tools] sdl2 --triplet x64-windows
```

This command takes about 10-30 minutes to complete, depending on your device.

> -   `curl` is optional, but required to read the serverlist, `curl[winssl]` is required to use the content store.

Reviewer 3: ```suggestion
-   `curl` is optional, but required to read the serverlist; `curl[winssl]` is required to use the content store.
```

This should be a semicolon. If we're fixing the docs, let's make them shine.

Reviewer 5: `curl` is strongly recommended. It's required for anything HTTP(S). The serverlist or content browser (or the built-in update checker) are just some "mainmenu" examples of this; the Lua HTTP API also needs curl to work.

(Frankly we might consider making curl mandatory in the future?)

Author: ```suggestion
-   `curl` is highly recommended, as it's required to read the serverlist; `curl[winssl]` is required to use the content store.
```

3. Select **Browse Build...** and select `DIR/minetest/build` (this is a new folder that CMake will prompt to create)
4. Select **Configure**
5. Choose the right Visual Studio version and target platform. Currently, Luanti uses Visual Studio 16 2019, but newer VS versions should work as well. The VS version has to match the version of the installed dependencies.
6. Choose **Specify toolchain file for cross-compiling**
7. Click **Next**
8. Select the vcpkg toolchain file e.g. `D:/vcpkg/scripts/buildsystems/vcpkg.cmake`
9. Click Finish
10. Wait until CMake generates the cache file (this may take about 10-30 minutes, depending on your device)
11. If there are any errors, solve them and hit **Configure**
12. Click **Generate**
13. Click **Open Project**
14. Compile Luanti inside Visual Studio.
    - If you get "Unable to start program '...\x64\Debug\ALL_BUILD'. Access is denied", try compiling via the CLI instead.

### b) CLI

Run the following script in PowerShell:

```powershell
cmake . -G"Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE=D:/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_CURSES=OFF
cmake --build . --config Release

> ```

Reviewer 3: We should not be encouraging in-tree builds. Also, I think CMake should automatically pick the correct generator if it is left unspecified.

Author: Hmm, personally I don't understand this, happy to take action if there are any suggestions :)

Reviewer 3: Running `cmake` from the build directory will cause CMake to output its build artifacts in the source tree. It's a pain to clean up, and it easily clutters output from tooling. It's better to run `cmake -B <name_of_build_directory>` which will limit the scope of CMake's shenanigans to single directory, which can be deleted in its entirely if it gets messed up later to start fresh. It also makes it easy to maintain multiple builds with different configurations.

Reviewer 3: I would suggest to change it to
```
cmake -B build -G"Visual Studio 16 2019" -DCMAKE_TOOLCHAIN_FILE=D:/vcpkg/scripts/buildsystems/vcpkg.cmake -DCMAKE_BUILD_TYPE=Release -DENABLE_CURSES=OFF
cmake --build build--config Release
```

I don't know how risky it would be to omit the `-G` option. It works on my machine because I use the CMake installed with the Visual Studio developer terminal, but I feel there's a risk that a different CMake installation might pick the wrong generator. I'm not well enough informed.

Author: I think we should keep the -G option for consistency

Author: However, users will need to update the value based on their installed version of VS, ref #15336

Author: Changed `.` to `build` :)

6. Choose **Specify toolchain file for cross-compiling**
7. Click **Next**
8. Select the vcpkg toolchain file e.g. `C:/vcpkg/scripts/buildsystems/vcpkg.cmake`
9. Click Finish
10. Wait until CMake generates the cache file (this may take about 10-30 minutes, depending on your device)
11. If there are any errors, solve them and hit **Configure**
12. Click **Generate**
13. Click **Open Project**

## Compile Luanti

There are two ways to compile Luanti: via Visual Studio or via CLI.

### Compile in Visual Studio

14. Compile Luanti inside Visual Studio.
    - If you get "Unable to start program '...\x64\Debug\ALL_BUILD'. Access is denied", try compiling via the CLI instead.

### Compile via CLI

While in the `path/to/minetest` folder, run the following script in PowerShell:

```powershell
vs="Visual Studio 16 2019" # or "Visual Studio 17 2022", whatever matches your system
toolchain_file="C:/vcpkg/scripts/buildsystems/vcpkg.cmake" # ensure this path is correct
cmake . -G $vs -DCMAKE_TOOLCHAIN_FILE=$toolchain_file -DCMAKE_BUILD_TYPE=Release -DENABLE_CURSES=OFF
cmake --build . --config Release
```

> ## Windows Installer using WiX Toolset

Author: Any advice on what to do with this "Windows Installer" section? I'm not sure what role it plays, or when it should come in. It mentions "compil[ing] binaries as above", but I don't remember seeing any "RUN_IN_PLACE" option

Reviewer 3: The `RUN_IN_PLACE` cache variable is used by Luanti to determine what paths to use for resources such as worlds and mods. I think I used the installer before, but it's been a while. We might have to rediscover how this works if no one remembers.

Author: Resolving for now as out-of-scope for this PR
